### PR TITLE
[Feature] Node rewards chain explorer. Resolves [#413]

### PR DESCRIPTION
**Pull request by apoorv-2204**
# Description

It adds a Reward chain explorer to UI, which includes ```:node_rewards and``` ```:mint_rewards``` tx types in chronological order.

Fixes 
- [ ]  #413 

## Type of change

Please delete options that are not relevant.

- Minor fix 
  - Name change:  get_gen_addr : genesis_address , wherever applicable.
  - Minor spec changes from number to non_negative_number for index.
  - Correct typos.
 
- New feature (non-breaking change which adds functionality)
   - New Reward explorer Ui.
   - New db method ```list_chain_addresses``` , takes genesis address and return a stream of address in old to recent order.
  
- This change requires a documentation update 
   - Brief Reward chain desc was given under ```/explorer/chain/rewards```  UI.

 - Review_changes
   - Transactions appear in page w.r.t previous page time reference.
   - Display pages according to nb of transactions.
   - ex: 40 transaction 4 page .
Screenshot_20220920_142749

- Deterministic test
   -  for faucet_controller_test.exs 

- [ ]  ####Future Improvements
   -  ets table with DE - queue
   -  Stream Transaction one by one from DB.
 
# How Has This Been Tested?

- Unit Test cases  were included wherever necessary
- Integration tests were included for live view
- DB tests.

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic/db.ex [hunk 1]
```diff
@@ -21,7 +21,7 @@ defmodule Archethic.DB do
   @callback write_transaction(Transaction.t()) :: :ok
   @callback write_beacon_summary(Summary.t()) :: :ok
   @callback write_transaction_chain(Enumerable.t()) :: :ok
-  @callback write_transaction(Transaction.t()) :: :ok
+  # @callback write_transaction(Transaction.t()) :: :ok
```
Comment (samuelmanzanera) [inline lib/archethic/db.ex:24]: Why did you remove this one ?

Comment (apoorv-2204) [reply lib/archethic/db.ex:24]: ```  @callback write_transaction(Transaction.t()) :: :ok```

**Comment by samuelmanzanera** [comment]
I got the previous button the first page

image

File: lib/archethic_web/templates/explorer/reward_chain_index.html.leex [hunk 2]
```diff
@@ -0,0 +1,56 @@
+<article class="message is-info">
+  <div class="message-header ">
+   <h1 class="subtitle is-size-4 heading has-text-white">Reward Chain</h2>
+  </div>
+  <div class="message-body">
+  Reward Chains are the basis for remuneration. During self-repair,  <strong>Archethic Rewards</strong> are created and distributed to Miners.
```
Comment (samuelmanzanera) [inline lib/archethic_web/templates/explorer/reward_chain_index.html.leex:6]: This is not really accurate, actually the rewards are not sent during self-repair, but the fees burnt are collected during self-repair. The rewards transactions are sent at a specific interval later, once per cycle.

**Comment by samuelmanzanera** [comment]
There is an issue with the pagination:

Page1 is ended with 2022/09/16 14:34:30
image


Page2: is started with 2022/09/16 14:39:30
image

**Comment by samuelmanzanera** [comment]
Looks like good!
I just have one remark, when we reach 10 and 10 on 2 pages, I have a third empty page. Best would be to not showing it.

File: lib/archethic_web/live/chains/reward_live.ex [hunk 3]
```diff
@@ -132,18 +132,24 @@ defmodule ArchethicWeb.RewardChainLive do
...
+          %{
+            address: addr,
+            type: (TransactionChain.get_transaction(addr, [:type]) |> elem(1)).type,
+            timestamp: DateTime.from_unix(timestamp, :millisecond) |> elem(1)
+          }
+        end)
+        |> Enum.reverse()
+
+      _ ->
+        [%{address: "", type: "", timestamp: ""}]
```
Comment (samuelmanzanera) [inline lib/archethic_web/live/chains/reward_live.ex:151]: Why are you not returning an empty list?
What is the unmatched case ? nil ?

**Comment by samuelmanzanera** [comment]
If I wait on the first page, the transactions are accumulated but the pagination is not activated. If I refresh after I will the see the pagination.
Maybe we should detect when we reach the 10txs, to display only the last 10 on the screen while activating the pagination.

File: lib/archethic_web/live/chains/reward_live.ex (hunk 3 above)
Comment (apoorv-2204) [reply lib/archethic_web/live/chains/reward_live.ex:144]: nice catch. True its nil

Comment (samuelmanzanera) [reply lib/archethic_web/live/chains/reward_live.ex:144]: Ok but why are you not displaying an empty list?

**Review by samuelmanzanera — APPROVED** [review]
Looks good

**Review by samuelmanzanera — APPROVED** [review]

**Comment by apoorv-2204** [comment]
Closing Comment:
we must leverage ets, queue for cache. Or may be idea from google/guava.
